docs/library/uasyncio.rst

```diff
@@ -0,0 +1,257 @@
+
+    This is a coroutine.
+
+class Task
+----------
+
+.. class:: Task()
+
+    This object wraps a coroutine into a running task.  Tasks can be waited on
+    using ``await task``, which will wait for the task to complete and return
+    the return value of the task.
+
+    Tasks should not be created directly, rather use `create_task` to create them.
+
+.. method:: Task.cancel()
+
+    Cancel the task by injecting a ``CancelledError`` into it.  The task may
+    or may not ignore this exception.
+
+class Event
+-----------
+
+.. class:: Event()
+
+    Create a new event which can be used to synchronise tasks.  Events start
+    in the cleared state.
+
+.. method:: Event.set()
+
+    Set the event.  Any tasks waiting on the event will be scheduled to run.
+
+.. method:: Event.clear()
+
+    Clear the event.
+
+.. method:: Event.wait()
+
+    Wait for the event to be set.  If the event is already set then it returns
+    immediately.
+
+    This is a coroutine.
+
+class Lock
+----------
+
+.. class:: Lock()
+
+    Create a new lock which can be used to coordinate tasks.  Locks start in
+    the unlocked state.
+
+    In addition to the methods below, locks can be used in an ``async with`` statement.
+
+.. method:: Lock.locked()
+
+    Returns ``True`` if the lock is locked, otherwise ``False``.
+
+.. method:: Lock.acquire()
+
+    Wait for the lock to be in the unlocked state and then lock it in an atomic
+    way.  Only one task can acquire the lock at any one time.
+
+    This is a coroutine.
+
+.. method:: Lock.release()
+
+    Release the lock.  If any tasks are waiting on the lock then the next one in the
+    queue is scheduled to run and the lock remains locked.  Otherwise, no tasks are
+    waiting an the lock becomes unlocked.
+
+TCP stream connections
+----------------------
+
+.. function:: open_connection(host, port)
+
+    Open a TCP connection to the given *host* and *port*.
+
+    Returns a pair of streams: a reader and a writer stream.
+
+    This is a coroutine.
+
+.. function:: start_server(callback, host, port, backlog=5)
+
+    Start a TCP server on the given *host* and *port*.  The *callback* will be
+    called with incoming, accepted connections, and be passed 2 arguments: reader
+    and writer streams for the connection.
+
+    Returns a `Server` object.
+
+    This is a coroutine.
+
+.. class:: Stream()
+    
+    This represents a TCP stream connection.  To minimise code this class implements
+    both a reader and a writer.  The names ``uasyncio.StreamReader`` and
+    ``uasyncio.StreamWriter`` are both aliases to this class.
+
+.. method:: Stream.get_extra_info(v)
+
+    Get extra information about the stream, given by *v*.  The valid values for *v* are:
+    ``peername``.
+
+.. method:: Stream.close()
+
+    Close the stream.
+
+.. method:: Stream.wait_closed()
+
+    Wait for the stream to close.
+
+    This is a coroutine.
+
+.. method:: Stream.read(n)
+
+    Read *n* bytes and return them.
```

Comment:
What are the exact semantics of read/write WRT "no short reads/writes"?

Reply:
Not quite sure...!  It should be exact, except if the connection has EOF.  That'll need to be clarified eventually.

Reply:
I checked on esp32: read always returns the full requested amt and `b''` once the socket is closed, i.e. EOF. Since the socket class "delegates" the stream read/write to the generic stream module, this is to be expected (for write as well).

Reply:
testing on the unix port, the read will return a short read if not all data is available, and this behaviour matches CPython's asyncio

Reply:
yup, on esp32 also returns short reads, grrr. would be nice to update the docs to say "up to *n* bytes"
does that mean writes can also be short?

Reply:
Little excerpt from http://docs.micropython.org/en/latest/library/uio.html:

> Whether there will be short operations or not depends on each particular class’ needs, but developers are strongly advised to favor no-short-operations behavior for the reasons stated above. For example, MicroPython sockets are guaranteed to avoid short read/writes. Actually, at this time, there is no example of a short-operations stream class in the core, [...]

Worth fixing Stream.read? Or change the above in the docs to start listing the exceptions? ;-)

Reply:
I updated the docs to specify that `read(n)` reads "up to n bytes".  This is how CPython works and how CPython's docs read.  Also, CPython implements `readexactly(n)` which we could also add.